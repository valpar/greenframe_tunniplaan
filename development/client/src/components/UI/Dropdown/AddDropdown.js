import { useState, useEffect } from "react";
import Select, { components } from "react-select";
import TooltipLarge from "../Tooltip/TooltipLarge";
import ConfirmModal from "../ConfirmModal/ConfirmModal";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleDown, faAngleUp } from "@fortawesome/free-solid-svg-icons";
import content from "../../../assets/content/content.json";

const Option = (props) => {
  const { icon, onEdit } = props.selectProps;

  const editHandler = () => {
    onEdit(props.data);
  };

  return (
    <>
      <components.Option {...props}>
        <div className="flex flex-row justify-between items-center">
          <div className="text-sm min-h-[2rem] leading-8 lg:h-auto lg:leading-7">
            {props.children}
          </div>

          <div
            className={`${
              props.children === "Lisa uus..." ? "hidden" : "flex"
            }`}
          >
            <div className="h-11/12 border-l border-borderGray" />
            <div
              className="group flex items-center justify-center w-12 h-12 lg:w-8 lg:h-8 ml-1 -mr-2 cursor-pointer rounded-full hover:bg-collegeGreen duration-150"
              onClick={editHandler}
            >
              <span className="group-hover:scale-105 duration-150">{icon}</span>
            </div>
          </div>
        </div>
      </components.Option>
    </>
  );
};

const AddDropdown = (props) => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const [showTooltip, setShowTooltip] = useState(false);
  const [menuIsOpen, setMenuIsOpen] = useState(undefined);
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [placeholderColor, setPlaceHolderColor] = useState("gray");
  const [editData, setEditData] = useState({});

  const { updateConfirmMessage } = content.confirmModalMessages;

  const icon = (
    <i
      aria-hidden="true"
      focusable="false"
      className="bi bi-pencil-fill inline-block"
    ></i>
  );

  const styles = {
    control: (baseStyles) => ({
      ...baseStyles,
      paddingLeft: "0.6rem",
      borderRadius: "none",
      minHeight: "2.8rem",
      boxShadow: "none",
      "&:hover": {
        border: "1px solid black",
        textColor: "black",
      },
      border: `${props.hasError ? "1px solid red" : "1px solid #CECECE"}`,
    }),
    placeholder: (baseStyles, state) => ({
      ...baseStyles,
      color: placeholderColor,
      marginLeft: "36px",
    }),
    indicatorsContainer: (baseStyles) => ({
      ...baseStyles,
      pointerEvents: "none",
    }),
    option: (baseStyles, state) => ({
      ...baseStyles,
      backgroundColor: state.isSelected
        ? "#E0E0E0"
        : state.isFocused
        ? "#E0E0E0"
        : "#fff",
      color:
        state.isClearable || state.isSelected
          ? "black"
          : state.isFocused
          ? "black"
          : "black",
      ":hover": {
        backgroundColor: state.isSelected ? "#E0E0E0" : "#E0E0E0",
      },
    }),
  };

  const changeHandler = (choice) => {
    let newArrayOfObj;
    if (props.isMulti) {
      newArrayOfObj = choice.map(({ value }) => ({
        [props.name + "Id"]: value,
      }));
    }
    if (!props.isMulti) {
      newArrayOfObj = [choice].map(({ value }) => ({
        [props.name + "Id"]: value,
      }));
    }

    if (newArrayOfObj.length > 0) {
      props.onChange(newArrayOfObj);
    } else {
      props.onChange([{ value: props.name }]);
    }
  };

  const inputChangeHandler = (e) => {
    if (props.onInputChange) {
      props.onInputChange(e);
    }
  };

  const mouseEnterHandler = () => {
    setShowTooltip(true);
    setPlaceHolderColor("black");
  };
  const mouseLeaveHandler = () => {
    setShowTooltip(false);
    setPlaceHolderColor("gray");
  };
  const declineHandler = () => {
    if (showConfirmModal) {
      setMenuIsOpen(undefined);
      return setShowConfirmModal(false);
    }
    props.onDecline(props.name);
  };
  const editHandler = (dataObj) => {
    setEditData(dataObj);
    setMenuIsOpen(false);
    setShowConfirmModal(true);
  };
  const confirmHandler = () => {
    setMenuIsOpen(undefined);
    setShowConfirmModal(false);
    props.onEdit({ ...editData, type: props.name });
  };
  const DropdownIndicator = (props) => {
    return (
      <components.DropdownIndicator {...props}>
        {!isMenuOpen && <FontAwesomeIcon icon={faAngleDown} className="px-1" />}
        {isMenuOpen && <FontAwesomeIcon icon={faAngleUp} className="px-1" />}
      </components.DropdownIndicator>
    );
  };

  const menuOpenHandler = () => {
    setIsMenuOpen(true);
  };
  const menuCloseHandler = () => {
    setIsMenuOpen(false);
  };

  return (
    <div
      onMouseEnter={mouseEnterHandler}
      onMouseLeave={mouseLeaveHandler}
      className="relative w-full h-auto"
      data-testid="add-dropdown"
    >
      {props.topLabel && <label>{props.topLabel}</label>}
      {props.onErrorMessage &&
        props.onErrorMessage !== "" &&
        (showTooltip || "") && (
          <div className="absolute -top-16 left-1/2 -translate-x-1/2 w-full max-w-xs">
            <TooltipLarge index={props.index} message={props.onErrorMessage} />
          </div>
        )}

      <Select
        onMenuClose={menuCloseHandler}
        onMenuOpen={menuOpenHandler}
        menuIsOpen={menuIsOpen}
        isSearchable
        icon={icon}
        onEdit={editHandler}
        components={{ DropdownIndicator, Option }}
        styles={styles}
        placeholder={props.label}
        options={props.options}
        onChange={changeHandler}
        isMulti={props.isMulti ? true : false}
        isClearable={false}
        onInputChange={inputChangeHandler}
        noOptionsMessage={(value) => (value = "")}
        value={
          props.value === ""
            ? props.value
            : props.options.filter(({ value }) => {
                if (typeof props.value === "number")
                  return value === props.value;
                return props.value?.find((v) => {
                  return Object.values(v)[0] === value;
                });
              })
        }
      />
      {(props.modalMessage?.show || showConfirmModal) && (
        <div className="absolute lg:top-16 lg:-left-6">
          <ConfirmModal
            modalMessage={
              showConfirmModal
                ? updateConfirmMessage
                : props.modalMessage?.message
            }
            topArrow={true}
            onConfirm={showConfirmModal ? confirmHandler : props.onConfirm}
            onDecline={declineHandler}
            modalButtons={props.modalButtons}
          />
        </div>
      )}
    </div>
  );
};

export default AddDropdown;
