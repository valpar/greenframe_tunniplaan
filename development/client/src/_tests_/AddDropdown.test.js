import React from 'react';
import { render, fireEvent, waitFor, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer'; //snapshot renderer
import AddDropdown from '../components/UI/Dropdown/AddDropdown';

const mockOptions = [{ value: 1, label: 'Option 1'}, { value: 2, label: 'Option 2' }];
const mockOnChange = jest.fn();
const mockOnInputChange = jest.fn();

afterEach(() => {
    cleanup();
});

describe('AddDropdown component', () => {
    // need to add options value (any) for a filter function in AddDropdown component to test rendering it
    it("if matches snapshot", () => {
        const tree = renderer.create(<AddDropdown options={mockOptions} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('renders the dropdown component', () => {
        render(<AddDropdown options={mockOptions}/>);
        expect(screen.getByRole('combobox')).toBeInTheDocument();
    });
    it('opens dropdown on click', () => {
        render(
          <AddDropdown
            name="test-dropdown"
            options={mockOptions} // Ensure mockOptions is defined, even if it's just an empty array or basic options
            onChange={mockOnChange}
            onInputChange={mockOnInputChange}
          />
        );
      
        // Try to open the dropdown
        const dropdown = screen.getByRole('combobox');
        fireEvent.mouseDown(dropdown);
      
        // Check if dropdown is in an expanded state
        expect(dropdown.getAttribute('aria-expanded')).toBe('true');
      });
});
