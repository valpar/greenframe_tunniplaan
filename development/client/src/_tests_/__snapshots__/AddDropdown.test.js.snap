// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDropdown component if matches snapshot 1`] = `
<div
  className="relative w-full h-auto"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className=" css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className=" css-1ok88yx-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-319lph-ValueContainer"
      >
        <div
          className=" css-rg4fjx-placeholder"
          id="react-select-2-placeholder"
        >
          Select...
        </div>
        <div
          className=" css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-2-placeholder"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className=""
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className=" css-q6enge-IndicatorsContainer"
      >
        <span
          className=" css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down px-1"
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
